openfe/setup/methods/rbfe_methods.py

```diff
@@ -0,0 +1,257 @@
+# This code is part of OpenFE and is licensed under the MIT license.
+# For details, see https://github.com/OpenFreeEnergy/openfe
+"""Contains RBFE methods
+
+This module implements.....
+
+"""
+from __future__ import annotations
+
+from openfe.setup import LigandAtomMapping, LigandMolecule
+from openfe.setup.methods import FEMethod
+from typing import Dict, Union
+from pydantic import BaseModel, validator
+from openff.units import unit
+
+# define a timestep
+# this isn't convertible to time (e.g ps) and is used to not confuse these two
+# definitions of "duration" within a simulation
+unit.define('timestep = [timestep] = _ = timesteps')
+
+
+class ForcefieldSettings(BaseModel):
+    """Settings describing the forcefield to use for each component
+
+    Attributes
+    ----------
+    forcefield
+      a mapping of each components name to the xml forcefield to apply
+    nonbonded_electrostatics
+      which nonbonded electrostatics method to use, currently only 'PME'
+      allowed
+    nonbonded_cutoff
+      range of nonbonded forces
+    constraints
+      which bonds to apply constraints to
+    rigid_water
+      whether apply rigid constraints to water molecules, default True
+    """
+    class Config:
+        arbitrary_types_allowed = True
+
+    # mapping of component name to forcefield path
+    forcefield: Dict[str, str]
+    nonbonded_electostatics = 'PME'
+    nonbonded_cutoff = 0.9 * unit.nanometer
+    constraints = 'HBonds'
+    rigid_water = True
+
+
+class StatePoint(BaseModel):
```

Comment:
I get the point of this, but maybe keeping it simple for users / developers would be better? Could we not sure have the one "config" class that just takes in everything you'd define for a simulation (the mdp to your FEP if you wish).

Reply:
So you'd flatten the entire collection of settings into one dict?

Reply:
If folks are meant to interact with it, yeah. Separation is great but it will become overwhelming quickly. Maybe we can cut it into two things - basic thermodynamic settings and alchemical settings? That way you have a clean separation and it's simple for folks to separate things out in their minds?

Reply:
I disagree with the complete flattening because it makes it impossible for users to plug in different combinations of methods, especially those that may be developed outside of OpenFE.

Reply:
Some degree of modularity doesn't hurt I guess, but I really wouldn't offer "statepoint" and things like "PME, cutoff, barostat" as separate settings, that makes no sense scientifically.

Let's think about this way, with the Langevin integrator, the temperature is a control variable of the integrator, which in turn is affected by the rate of collision. It becomes convoluted really quickly.

Reply:
If we really want to keep things separate (and we ignore some issues like integrator == thermostat), then split it in a more logical manner imho:

Integrator settings + Reporter settings + Thermostat settings + Barostat settings + Alchemical settings

Essentially unless there's a real reason for it I'd just axe StatePoint.

Reply:
> Some degree of modularity doesn't hurt I guess, but I really wouldn't offer "statepoint" and things like "PME, cutoff, barostat" as separate settings, that makes no sense scientifically.

Agreed. I think we should mimic the original API as closely as possible. I'd do a set of setting for the `openmm.app.ForceField` object and its `createSystem` method. I'd just use the OpenMM integrator and barostat directly (they can be serialized to XML). Another set of settings for the `HybridTopology`. Another set for the `openmm.app.Simulation` object (need a place to specify platform settings). Another set to describe the solvation process (basically wrapping `addSolvent` and possible some associated things). Finally, some global settings that collects all these individual settings. Maybe I'm missing another touch point specific to the alchemical setup?

Reply:
> Essentially unless there's a real reason for it I'd just axe StatePoint.

+1. I think it's an attempt to avoid having to repeat the temperature setting in the barostat. I think we can just use the barostat directly.

Reply:
So the question really is here, how much should we be buying into the OpenMM ecosystem?

I would suggest keeping things relatively abstracted from the eventual objects that use these settings if we want to use FEMethod for multiple MD engines eventually.

Reply:
I think that settings are going to be engine-specific. We can't assume we'll have the same integrators, etc. available everywhere. Because of that, I say stay as similar to the engine as possible.

IMO, all the engine settings for Gromacs should just be taken from an mdp file. Go to the users where they are; don't make a new whole new thing they need to learn to use our stuff.

Reply:
I guess it depends at what point in the abstraction layer we're looking at tbh. If it is OpenMM only then we should reflect that in the method name.

If it was engine non-sepcific; every engine we use will have the same integrator type (if it's not Langevin it's not going to work as an FEP). Barostats are more or a pain but we could just fail if we don't have access to the right thing?

re: gmx I'm not sure I agree with mdp file as the entry point, we should be able to have similar entry points for every method, and making mdp files from mdp files just seems like a terrible deal for users. But anyways that's not an issue to be fixed now.

Reply:
> I guess it depends at what point in the abstraction layer we're looking at tbh. If it is OpenMM only then we should reflect that in the method name.

I would say put them in a module to reflect that, rather than change the class names. `openfe.setup.openmm_settings`. You're rarely going to use more than one engine at once (it is possible, but not a typical use case.)

> If it was engine non-sepcific; every engine we use will have the same integrator type (if it's not Langevin it's not going to work as an FEP). Barostats are more or a pain but we could just fail if we don't have access to the right thing?

What about solvation? That's something that different engines do differently as well, and the algorithms are going to be different enough that you probably can't count on the same result from two engines. (Granted, you can solvate with Gromacs and run with OpenMM, but only if you have both installed on your system.) Note that we have to specify *which* Langevin as well, and not all are supported by all engines. The problem is that as we add engines, the number things that are "this only exists for this engine" grows. If we want to support a wide variety of feautres, the number of things that are engine-specific are going to be larger than the number of things that are shared between engines. At some point we need to maintain a database of what works with what engine (and what version of the engine). 

Aside: Why the need for Langevin? (Point to reference?) I would at least think that other stochastic integrators would work.

Reply:
Another point: what if an engine supports different options in the "force field" term? Example, maybe some method for electrostatics is/is not supported. If the settings are engine-specific, we can validate this on creation. If not, then it fails later.

My assumption is that most people have decided on the engine they will use before they start setting up the simulation. (In fact, they need to, to select an integrator/barostat that is supported.)

Reply:
> Run (or reporter) settings

In an ideal world things like "Run" settings should be part of the simulation plan (or whatever Richard named it). I guess we can't current affect reporter settings too well, but you could consider a case where someone goes "o yeah I want to run a 5 ns equilibration printing every 100ps, but I was a 10 ns production printing every 2 ps".

Reply:
> In an ideal world things like "Run" settings should be part of the simulation plan (or whatever Richard named it).

This is an interesting point. Does the fact that it is the only thing that *isn't* backend-specific kind of suggest that it should be pushed back to a later point? Probably yes, especially if you're thinking about using some adaptive sampling, where the decision made now becomes irrelevant.